website_canonical_url/models/website.py

```diff
@@ -0,0 +1,35 @@
+# -*- coding: utf-8 -*-
+# Copyright initOS GmbH 2016
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import models, api
+from odoo.http import request
+from urlparse import urlparse, urlunparse
+
+
+class Website(models.Model):
+    _inherit = 'website'
+
+    @api.multi
+    def get_canonical_url(self, req=None):
+        canonical_url = None
+        if req is None:  # pragma: no cover
+            req = request
+        if req and req.httprequest.path:
+            lang = self.env.lang
+            if lang == request.website.default_lang_code:
+                canonical_url = req.httprequest.path
+            else:
+                url_parts = urlparse(req.httprequest.path)
+                url_parts = list(url_parts)
+                # change the path of the url
+                url_parts[2] = lang + url_parts[2]
+                canonical_url = urlunparse(url_parts)
+        # Special case for rerouted requests to root path
+        try:
+            if (canonical_url.startswith("/page/") and
+                    self.menu_id.child_id[0].url == canonical_url):
+                canonical_url = "/"
+        except:  # pragma: no cover
+            pass
+        return canonical_url
```

Comment:
This method should return the protocol and hostname too.

What I'm not sure about is if it should get them from the request or from `web.base.url` config param. I think it should be from the param, since that's something you can configure, but then it could be different by website.

Possibly the best solution would be to add a new field to the `website` model, called `canonical_domain`, where one could define the full protocol+hostname for all canonical urls that come from this website.

Then, here, we'd just `return urlparse.urljoin(website.canonical_domain, canonical_url)`

Reply:
Totally agree on the abs url. Regarding where to take the domain:
the doc you linked clearly states that we should have 1 single URL no matter the subdomain you might use (like m.mydomain.com for mobile). So, to me the best solution seems to use `web.base.url`.

> but then it could be different by website.

@Yajo I don't get you here: that param is one and only one for one DB thus is going to be the same. So unless you want to serve contents from different instances/DBs the param is going to be always the same. The issue might come when you have multiple websites on the same DB that are not related to each other and don't share the same domain. In this case, yes, probably having the canonical domain on the website object is a good idea.

Am I missing anything?

Reply:
> The issue might come when you have multiple websites on the same DB that are not related to each other and don't share the same domain.

You nailed it. `website_multi_theme` addon comes to my mind. It would be affected with this problem and it would be incompatible with `website_canonical_url` then.

Another good thing of having it per website is that several websites could share the same canonical domain, which is actually why the recommendation recommends such thing.

`web.base.url` could be a fallback value if the canonical domain is undefined.

Reply:
Ok, then I'll go for new field + fallback to param. Not sure about the protocol tho, since web.base.url usually does not specify it. For this we might want to stick to the request protocol which should be ok if you setup your web server properly. No?

Reply:
Hmmmm AFAICR `web.base.url` always includes the protocol, and the new canonical_domain field should be constrained to force including it always I think.

Reply:
I came to the same conclusion : it would be better with a new field. But I thought it wouldn't be accepted. I found the solution to call the canonical domain in the template file, because it would be easier to adapt the template to each website.